Diátaxis: "first recipe" tutorial (#2705)
conda-forge/conda-forge.github.io: docs/tutorials/first-recipe.md

---
tags: [tutorial]
---

# Generating and building your first recipe

In this tutorial, we will generate your first recipe for a pure Python package, test it by building
the package and use a linter to verify that it follows the conda-forge best practices.

The tutorial assumes that you have a working conda installation, and that you are running Linux or
macOS. In case of Linux, you also need to have a working Docker installation.

## Prepare the development environment

While it is largely possible to work on conda-forge packages using system tools, to provide
consistency we will assume here that all the needed tools are installed in a conda environment.

Start a terminal with your conda installation activated. In that terminal, create a new environment
with the needed packages, then activate it:

```bash
conda create -n conda-forge-dev -y conda-smithy git grayskull python

> conda activate conda-forge-dev

Reviewer: It may be just me, but I'm not a huge fan of installing git through conda. git to me is a global tool that shouldn't be per-environment (`pixi global` works great for this). How about
```suggestion
conda create -n conda-forge-dev -y conda-smithy grayskull python
conda activate conda-forge-dev
conda install -y git  # if you do not have a working `git` installation on your system yet
```

## Clone the staged-recipes repository

The [staged-recipes](https://github.com/conda-forge/staged-recipes) repository is where new recipes
for conda-forge are contributed. While we aren't going to be contributing a new recipe, it has a
convenient infrastructure for testing one.

Enter a directory of your choice, clone the repository and then enter it:

```bash
git clone --depth=1 https://github.com/conda-forge/staged-recipes
cd staged-recipes
```

We are using `--depth=1` to reduce the download size and disk space consumption, as we do not need
the repository's history.

## Generate the initial recipe and build it

Let's try to create a recipe for [pylast 7.0.1](https://pypi.org/project/pylast/7.0.1/). Since it
is a PyPI package, we can start with an automatically generated recipe. We'll need to:

1. Create a git branch for our work, so that the `main` branch remains intact.
2. Run `grayskull` to create the recipe.

Type in the terminal:

```bash
git checkout -b pylast
grayskull pypi --use-v1-format --strict-conda-forge -o recipes pylast==7.0.1
```

Here's a quick wraparound of the arguments passed:

- `pypi` indicates that a recipe should be generated for a package from PyPI.

> - `--use-v1-format` uses the modern v1 recipe format.

Reviewer: "modern" has a way of becoming stale quickly. How about
```suggestion
- `--use-v1-format` uses v1 recipe format (v0 is the `meta.yaml` format, now considered legacy).
```

- `--strict-conda-forge` follows conda-forge rules more strictly.
- `-o recipes` causes the recipe to be output into a subdirectory of `recipes` directory.
- `pylast==7.0.1` specifies the package name and version. Normally, you'd only specify a package
  name, but we're using a specific version here to make the results more predictable.

This command will write a `recipes/pylast/recipe.yaml` file. This file is the recipe, containing the
instructions for building a package. It is already working, and we can use the `build-locally.py`
script to build the package:

```bash
python build-locally.py
```

The script will ask you to select the config appropriate for the system you're running. Once you
type the correct number and push enter, the script will create a build environment and build the
package verbosely. Once the build completes, the new package will be found in
`build_artifacts/noarch` directory, e.g.:

```
-rw------- 1 user user  36K 01-28 18:57 pylast-7.0.1-pyhcb3b5b4_0.conda

> ```

Author: I've already replied to the top comment before reading this. It being more likely to fail than other steps is indeed a good point. However, I still believe that actually building the package makes all the difference between packaging work and just copy-pasting a text file.

instructions for building a package. It is already working, and we can use the `build-locally.py`
script to build the package:

```bash
python build-locally.py
```

The script will ask you to select the config appropriate for the system you're running. Once you
type the correct number and push enter, the script will create a build environment and build the
package verbosely. Once the build completes, the new package will be found in
`build_artifacts/noarch` directory, e.g.:

```
-rw------- 1 user user  35K 01-01 14:00 pylast-7.0.1-pyh1d6dcf3_0.conda
```

## Look at the recipe file

Now is a good opportunity to take a look at the recipe file we've obtained. Open
`recipes/pylast/recipe.yaml` in your favorite text editor. We will go over the YAML file section by
section:

```yaml
schema_version: 1
```

This preamble indicates that this is a v1 recipe. It will always be present on top of the v1 recipe
files.

```yaml
context:
  name: pylast
  version: 7.0.1
```

This section defines helper variables that can be used elsewhere in the recipe. Here, `name` and
`version` are declared, so that we can update them in one place, and reference elsewhere in the

> recipe.

Reviewer: Here we run into the kind of friction that got [discussed](https://conda-forge.zulipchat.com/#narrow/channel/457337-general/topic/.E2.9C.94.20Motivation.20for.20.60name.60.20field.20in.20.60context.60.20for.20recipes.3F/with/567023367) just hours ago in zulip: the `name:` variable is pointless, and should be avoided; at the same time, this is the kind of recipe that `grayskull` currently generates. I'm fine to continue describing the status quo like this, but just wanted to make you aware of the (apparently) upcoming RFC to change this.

Author: Yeah, I also found it weird that it includes it, but obviously the tutorial needs to actually refer to what users are going to get. Ofc if this changes, I'll update the PR, or file a new one if it's merged.

Author: As in comment in the actual code sample, telling the reader that things will change?

  sha256: 319251236ba5c3e907232aacf1d6a7ff831f2243e85ace6ec6623a552ec2e0eb
```

This section specifies how to download the source distribution to build. It specifies the URL and a
SHA256 hash that is used to verify its authenticity. Again, variables are used to replace some
repeating bits.

```yaml
build:
  number: 0
  noarch: python
  script: ${{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
```

Here, three important bits of information are specified:

1. The build number, which is a number indicating the recipe revision. Whenever you make major
   changes to a recipe that justify building new packages, you increment it.
2. The `noarch` key that indicates that we're building a pure Python package, and so we do not need
   to build separate packages for different conda-forge platforms.
3. The actual build command invocation. We're using the Python package manager `pip` here to build
   and install our package. We are passing `--no-deps` to avoid automatically installing its
   dependencies, and `--no-build-isolation` to make the build use our build environment rather than
   creating a new one.

```yaml
requirements:
  host:

>     - python >=3.10

Reviewer: Is it an intentional choice here to not use `${{ python_min }}`. I'm fine for the "first recipe" to skip that, but wanted to double-check

dependencies are needed to build the package, while `run` dependencies are needed to run it once
it's installed.

```yaml
tests:
  - python:
      imports:
        - pylast
      pip_check: true
```

This section specifies how to test the package. In this case, it performs the absolute minimum
testing expected of Python package: checking if `pylast` can be imported successfully, and running
`pip check` to verify that the needed dependencies were installed.

```yaml
about:
  summary: A Python interface to Last.fm and Libre.fm
  license: Apache-2.0
  license_file:
    - LICENSE.txt
    - COPYING
  repository: https://github.com/pylast/pylast
```

The `about` section provides additional package information. Notably, it includes a short package
description, its license information along with a list of files containing the license inside the
source archive, and the repository URL.

```yaml
extra:
  recipe-maintainers:
    - your-name
```

Finally, the `extra` section specifies the package maintainers. It will normally contain your user
name. If it's different from your GitHub username, you will want to update it before submitting a

> recipe.

Reviewer: I don't understand where you're going with this. It's never people's real name. Only GH user handles

```suggestion
Finally, the `extra` section specifies the package maintainers. It will normally contain your user
Github user name.
```

> ## Run the test suite

Reviewer: I feel everything past this point is already a follow-up article. Let's ensure that the "first recipe" is a friendly looking (also in terms of overall length) article, to not scare people unduly 😅

Author: Well, I felt that a "first recipe" should actually involve some work beyond starting the generator.
